Create ScreenRayTracker.cs

JAQuent opened the pull request.

JAQuent commented:
In additition to the screen tracker I also quickly edited the CSV session builder and created a TSV session builder with it because people might want to use "," in their strings.

jackbrookes commented:
Hi, I don't have time to review right now but I can see several problems immediately, that would need to be addressed before merging. I will add comments.

jackbrookes commented on Assets/UXF/Scripts/Trackers/ScreenRayTracker.cs:
```diff
@@ -0,0 +1,150 @@
+    	numRays = y.Count;
+
+    	// Create var to reset the variable
+    	List<string> nameOfObjects = new List<string>();
```
better to create a field, and clear it, rather than make a new list each time

jackbrookes commented on Assets/UXF/Scripts/Trackers/ScreenRayTracker.cs:
```diff
@@ -0,0 +1,150 @@
+    /// <summary>
+    /// Function to detect objects on screen by rays
+    /// </summary>
+    List<string> ray2detectObjects(List<float> x, List<float> y, Camera cam){
+    	// Get number of rays
+    	numRays = y.Count;
```
what if x & y are different lengths?

jackbrookes commented on Assets/UXF/Scripts/Trackers/ScreenRayTracker.cs:
```diff
@@ -0,0 +1,150 @@
+    /// <summary>
+    /// Function to detect objects on screen by rays
+    /// </summary>
+    List<string> ray2detectObjects(List<float> x, List<float> y, Camera cam){
```
Methods should be PascalCase

jackbrookes commented on Assets/UXF/Scripts/Trackers/ScreenRayTracker.cs:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    /// <summary>
+    /// Function to detect objects on screen by rays
```
comment is not useful

jackbrookes commented on Assets/UXF/Scripts/Trackers/ScreenRayTracker.cs:
```diff
@@ -0,0 +1,150 @@
+    public override string MeasurementDescriptor => "ObjectsOnScreenTracker";
+    public override IEnumerable<string> CustomHeader => new string[] { "rayIndex", "x", "y", "objectDetected"};
+
+    // Get values
```
comment is not useful

jackbrookes commented on Assets/UXF/Scripts/Trackers/ScreenRayTracker.cs:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    /// <summary>
+    /// Stops the recording. This method needs to be added to [UXF_Rig] events called On Trial End
```
why does it need to be added to `On Trial End`? trackers have start and stop recording methods

jackbrookes commented on Assets/UXF/Scripts/Trackers/ScreenRayTracker.cs:
```diff
@@ -0,0 +1,150 @@
+    public override IEnumerable<string> CustomHeader => new string[] { "rayIndex", "x", "y", "objectDetected"};
+
+    // Get values
+    protected override UXFDataRow GetCurrentValues(){
+        return currentRow;
+    }
```
this means we will end up recording the values twice?

JAQuent replied:
Mhmm, strange in the result files each row is unique as far as I can see.

jackbrookes replied:
what if you use `UpdateMode` that is not manual?

JAQuent commented on Assets/UXF/Scripts/Etc/UXFDataTable.cs:
```diff
@@ -73,6 +73,38 @@ public static UXFDataTable FromCSV(string[] csvLines)
             return table;
         }
 
+        /// <summary>
+        /// Build a table from lines of TSV text.
+        /// </summary>
+        /// <param name="tsvLines"></param>
+        /// <returns></returns>
+        public static UXFDataTable FromTSV(string[] tsvLines)
+        {
+            string[] headers = tsvLines[0].Split('\t');
+            var table = new UXFDataTable(tsvLines.Length - 1, headers);
+
+            // traverse down rows
+            for (int i = 1; i < tsvLines.Length; i++)
+            {
+                string[] values = tsvLines[i].Split('\t');
+
+                // if last line, just 1 item in the row, and it is blank, then ignore it
+                if (i == tsvLines.Length - 1 && values.Length == 1 && values[0].Trim() == string.Empty ) break;
+
+                // check if number of columns is correct
+                if (values.Length != headers.Length) throw new Exception($"TSV line {i} has {values.Length} columns, but expected {headers.Length}");
+
+                // build across the row
+                var row = new UXFDataRow();
+                for (int j = 0; j < values.Length; j++)
+                    row.Add((headers[j], values[j].Trim('\"')));
+
+                table.AddCompleteRow(row);
+            }
+
+            return table;
+        }
```
Completely agree :)

JAQuent replied:
Also, just changing the experimenter builder is not enough because then what might happen is that a variable such as a string that is to be displayed to the participant contains and a comma, which UXF then tries to save, which doesn't work if the the output file is an .csv. I currently have work-around where I just replace a comma with an underscore before it is saved but tihis prevents me to copy the data from the input file to the result file.

JAQuent commented:
Hey Jack, I've finally got around addressing your suggestions about the ray tracker because I never got notification of those until you closed the issue. Is there still point working on this? I am actually starting data collection for new project soon so I wanted to tidy all the code up including this. I've two questions: a) Why are the rows recorded twice. I think I might misunderstand how this might work. b) The reason I add the stop method On Trial Begin/End is because I think it only works if the update type is set to manual and I was under the impression that I have to do this like this then. Is that not true?

jackbrookes commented:
> a) Why are the rows recorded twice

The trackers have an updateType which can be LateUpdate, FixedUpdate, Manual. If we don't select manual, there will be multiple times RecordRow is called per frame

> b) The reason I add the stop method On Trial Begin/End is because I think it only works if the update type is set to manual and I was under the impression that I have to do this like this then. Is that not true?

Regardless of UpdateType (which should not be needed to be set to manual) the UXF trial automatically calls the StopRecording method at the end of a trial - no need to hook into events

JAQuent commented:
Hey, so when I don't attach the start and stop methods to the On Trial Begin/End then a .csv tracking file is saved but it is empty.  Could that be because I use version 2.4.2?

As to your other question, I used manual because I wanted to be able to record mutliple rows per frames. See here https://github.com/immersivecognition/unity-experiment-framework/issues/91.

jackbrookes commented:
> Hey, so when I don't attach the start and stop methods to the On Trial Begin/End then a .csv tracking file is saved but it is empty. Could that be because I use version 2.4.2?

Yeah you would need to also restore the functionality to allow `LateUpdate` and `FixedUpdate` `updateType`s to work

> As to your other question, I used manual because I wanted to be able to record mutliple rows per frames. See here https://github.com/immersivecognition/unity-experiment-framework/issues/91.

Yeah I see - I think this is confusing because now the `updateType` must be set to manual. We would need a nicer way for this to work for it to be a part of UXF. It may be better to override the `RecordRow()` method for example

JAQuent commented:
Do you have an idea how to do this? I happy to link into this. 

Not sure how this will effect other things but would it be possible to simply take away the option fo the other update types and force manual here?
